tests/src/test/scala/scalaz/ReaderWriterStateTTest.scala

```diff
@@ -27,12 +27,19 @@ object ReaderWriterStateTTest extends SpecLite {
 
   checkAll(bindRec.laws[RWSOptInt])
   checkAll(monadPlus.strongLaws[RWSOptInt])
-
+  
+  "ReaderWriterStateT can be trampolined without stack overflow" in {
+    import scalaz.Free._
+    val result = (0 to 10000).toList.map(ii => ReaderWriterStateT[Trampoline, Unit, String, Int, Int]((_, i: Int) => Trampoline.done(("", i, ii))))
+      .foldLeft(ReaderWriterStateT[Trampoline, Unit, String, Int, Int]((_, i: Int) => Trampoline.done(("", i, i))))( (a, b) => a.flatMap(_ => b))
+    10000 must_=== result.run((),0).run._3
+  }
+ 
   object instances {
     def functor[F[_]: Functor, R, W, S] = Functor[RWST[F, R, W, S, ?]]
     def plus[F[_]: Plus, R, W, S1, S2] = Plus[IRWST[F, R, W, S1, S2, ?]]
     def plusEmpty[F[_]: PlusEmpty, R, W, S1, S2] = PlusEmpty[IRWST[F, R, W, S1, S2, ?]]
-    def bindRec[F[_]: BindRec : Applicative, R, W: Semigroup, S] = BindRec[RWST[F, R, W, S, ?]]
+    def bindRec[F[_]: BindRec : Monad, R, W: Semigroup, S] = BindRec[RWST[F, R, W, S, ?]]
```

Comment:
is it necessary to change this? could it be done with just existing BindRec? It seems like there is maybe a problem with BindRec that it couldn't be done with that and needs an explicit `Trampoline` but I'm probably missing something. cc @puffnfresh